File: Agent项目案例/model/factory.py
from abc import ABC, abstractmethod
from typing import Optional

from langchain_community.chat_models import ChatTongyi
from langchain_community.embeddings import DashScopeEmbeddings
from langchain_core.embeddings import Embeddings
from langchain_core.language_models import BaseChatModel

from agent项目案例.utils.config_handler import rag_conf
import os

# 模型工厂
class BaseModelFactory(ABC):
    @abstractmethod
    def generator(self) -> Optional[Embeddings | BaseChatModel]:
        pass

# 聊天模型工厂
class ChatModelFactory(BaseModelFactory):
    def generator(self) -> Optional[Embeddings | BaseChatModel]:
        api_key = rag_conf.get("api_key") or os.getenv("DASHSCOPE_API_KEY")
        return ChatTongyi(model=rag_conf["chat_model_name"], api_key=api_key)

# 嵌入模型工厂
class EmbeddingFactory(BaseModelFactory):
    def generator(self) -> Optional[Embeddings | BaseChatModel]:
        api_key = rag_conf.get("api_key") or os.getenv("DASHSCOPE_API_KEY")
        return DashScopeEmbeddings(model=rag_conf['embedding_model_name'], dashscope_api_key=api_key)

chat_model = ChatModelFactory().generator()
embed_model = EmbeddingFactory().generator()

File: agent_next/src/agent/model/factory.ts
import { BaseChatModel } from "@langchain/core/language_models/chat_models";
import {Embeddings} from "@langchain/core/embeddings";
import { ChatAlibabaTongyi } from "@langchain/community/chat_models/alibaba_tongyi";
import { AlibabaTongyiEmbeddings } from "@langchain/community/embeddings/alibaba_tongyi";
import ragConfig from "@/agent/config/rag";
import {ChatOpenAI} from "@langchain/openai";

// 抽象工厂基类（对应 ABC 抽象类）
export abstract class BaseModelFactory {
  abstract generator(): Embeddings | BaseChatModel | null;
}

// 通义千问聊天模型工厂
export class ChatModelFactory extends BaseModelFactory {
  generator(): BaseChatModel {
    const apiKey = process.env.DASHSCOPE_API_KEY;
    if (!apiKey) {
      throw new Error('DASHSCOPE_API_KEY 环境变量未配置');
    }
    return new ChatOpenAI({
      model: ragConfig.chat_model_name,
      streaming: true,
      temperature: 0.7,
      openAIApiKey: apiKey,
      configuration: {
        baseURL: "https://dashscope.aliyuncs.com/compatible-mode/v1",
      },
    });
  }
}

// Ollama 聊天模型工厂
// export class ChatOllamaModelFactory extends BaseModelFactory {
//   generator(): BaseChatModel {
//     return new ChatOllama({
//       model: ragConf.chatOllModelName,
//       streaming: true,
//     });
//   }
// }

// 嵌入模型工厂
export class EmbeddingFactory extends BaseModelFactory {
  generator(): Embeddings {
    const apiKey = process.env.DASHSCOPE_API_KEY;
    
    if (!apiKey) {
      throw new Error('DASHSCOPE_API_KEY 环境变量未配置');
    }
    
    return new AlibabaTongyiEmbeddings({
      modelName: ragConfig.embedding_model_name,
      apiKey: apiKey,
    });
  }
}

export const ChatAlibabaTongyiModel = new ChatModelFactory().generator();
// export const chatOllModel = new ChatOllamaModelFactory().generator();
export const embedModel = new EmbeddingFactory().generator();